### PR TITLE
Vulners Plugins -> Develop branch

### PR DESCRIPTION
**Pull request by vulnersCom**
Hi again!
I've fixed all issues figured by you.
Please, take a look :)
Cheers!

### CODE REVIEW

File: w3af/plugins/grep/vulners_scanner.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        return self._vulners_api
+
+    def check_vulners(self, software_name, software_version, check_type):
+        vulnerabilities = {}
+        cached_result = self._vulnerability_cache.get((software_name, software_version, check_type))
+        if cached_result:
+            return cached_result
+        # Ask Vulners about vulnerabilities
+        if check_type == 'software':
+            vulnerabilities = self.get_vulners_api().softwareVulnerabilities(software_name, software_version)
```
Comment (andresriancho) [inline w3af/plugins/grep/vulners_scanner.py]: What happens if the user entered an invalid API key? Will this line raise an exception?

File: w3af/plugins/grep/vulners_scanner.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        elif check_type == 'cpe':
+            cpe_string = "%s:%s" % (software_name, software_version)
+            vulnerabilities = self.get_vulners_api().cpeVulnerabilities(cpe_string.encode())
```
Comment (andresriancho) [inline w3af/plugins/grep/vulners_scanner.py]: Same potential issue here

File: w3af/plugins/grep/vulners_scanner.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    def __init__(self):
+        GrepPlugin.__init__(self)
+        # Vulners Rules JSON url
+        self._vulners_rules_url = URL('https://raw.githubusercontent.com/vulnersCom/detect-rules/master/rules.json')
+
+        # Vulners shared objects
+        self._vulners_api = None
+        self._vulners_api_key = None
+        self.rules_table = None
+        self.rules_updated = False
```
Comment (andresriancho) [inline w3af/plugins/grep/vulners_scanner.py]: `self._multi_re` should be defined as None here too.

**Comment by andresriancho** [comment]
Added only a few small comments, after that we're ready to merge :+1:

File: w3af/plugins/grep/vulners_scanner.py (hunk 1 above)
Comment (vulnersCom) [reply w3af/plugins/grep/vulners_scanner.py]: It will. Yep, good idea to refactor this moment.

File: w3af/plugins/grep/vulners_scanner.py (hunk 3 above)
Comment (vulnersCom) [reply w3af/plugins/grep/vulners_scanner.py]: Got it.

**Comment by vulnersCom** [comment]
Updated.
Looks like we will need try/excepting every call to the API.
Not to stop plugin if single error happened.

**Comment by andresriancho** [comment]
Done! Thanks for all the changes :+1:

**Comment by andresriancho** [comment]
@vulnersCom I'm trying to write a unittest for this plugin. Could you please give me an example of a vulnerable HTTP response?

**Comment by andresriancho** [comment]
PS: Just applied many cosmetic changes to the plugin https://github.com/andresriancho/w3af/commit/de8bba5506b723ba2b29eeca596fa636a3e6c56d

**Comment by vulnersCom** [comment]
Great! Thank's for your support!
For unit testing I used this website: http://ciur.ru/vtk/default.aspx (oooops).
It's producing this output:
```
w3af>>> start
Buffer overflow in Microsoft Internet Information Services (IIS) 7.5, when FastCGI is enabled, allows remote attackers to execute arbitrary code via crafted headers in a request, aka "Request Header Buffer Overflow Vulnerability.".  This vulnerability was found in the request with id 19.
Unspecified vulnerability in Microsoft IIS 6.0, 7.0, and 7.5, when Extended Protection for Authentication is enabled, allows remote authenticated users to execute arbitrary code via unknown vectors related to "token checking" that trigger memory corruption, aka "IIS Authentication Memory Corruption Vulnerability.".  This vulnerability was found in the request with id 19.
Heap-based buffer overflow in the TELNET_STREAM_CONTEXT::OnSendData function in ftpsvc.dll in Microsoft FTP Service 7.0 and 7.5 for Internet Information Services (IIS) 7.0, and IIS 7.5, allows remote attackers to execute arbitrary code or cause a denial of service (daemon crash) via a crafted FTP command, aka "IIS FTP Service Heap Buffer Overrun Vulnerability." NOTE: some of these details are obtained from third party information. This vulnerability was found in the request with id 19.
Microsoft Internet Information Services (IIS) 7.5 uses weak permissions for the Operational log, which allows local users to discover credentials by reading this file, aka "Password Disclosure Vulnerability.".  This vulnerability was found in the request with id 19.
Stack consumption vulnerability in the ASP implementation in Microsoft Internet Information Services (IIS) 5.1, 6.0, 7.0, and 7.5 allows remote attackers to cause a denial of service (daemon outage) via a crafted request, related to asp.dll, aka "IIS Repeated Parameter Request Denial of Service Vulnerability.".  This vulnerability was found in the request with id 19.
Scan finished in 11 seconds.
Stopping the core...
w3af>>>
```

**Comment by andresriancho** [comment]
Added unittest https://github.com/andresriancho/w3af/commit/e0f12d4e60ddf253719e190038642ce8fcd3f977

**Pull request merged**